src/main/java/jp/ac/utokyo/rcast/karkinos/filter/TerminalMismatch.java

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright Hiroki Ueda
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package jp.ac.utokyo.rcast.karkinos.filter;
+
+import jp.ac.utokyo.rcast.karkinos.utils.TwoBitGenomeReader;
+
+import htsjdk.samtools.Cigar;
+import htsjdk.samtools.CigarElement;
+import htsjdk.samtools.CigarOperator;
+import htsjdk.samtools.SAMRecord;
+
+import java.util.List;
+
+
+public class TerminalMismatch {
+
+    //add 2020/12/17 for FFPE anneling near repeat, extends soft clipping
+    public static int terminalMismatch(SAMRecord sam, TwoBitGenomeReader tgr,int extraCheckLen) {
+        try {
+            return Math.max(leftMismatch(sam,tgr,extraCheckLen),rightMismatch(sam,tgr,extraCheckLen));
+        }catch(Exception ex){
+            //could not extend soft clips
+            ex.printStackTrace();
+        }
+        return 0;
+    }
+
+    private static int leftMismatch(SAMRecord sam, TwoBitGenomeReader tgr, int extraCheckLen) {
+        int alignmentStart = sam.getAlignmentStart();
+        int miscount = 0;
+
+        for(int n=alignmentStart;n<=alignmentStart+extraCheckLen;n++){
+            //
+            char refNuc = tgr._getGenomeNuc(n,true);
+            int idx = sam.getReadPositionAtReferencePosition(n)-1;
+            if(idx>=0 && idx<sam.getReadLength()){
+                char readNuc = sam.getReadString().charAt(idx);
+                if(!equalnuc(refNuc,readNuc)){
+                    miscount++;
+                }
+            }
+        }
+
+        return miscount;
+    }
+
+    private static int rightMismatch(SAMRecord sam, TwoBitGenomeReader tgr, int extraCheckLen) {
+        int alignmentEnd = sam.getAlignmentEnd();
+        int miscount = 0;
+
+        for(int n=alignmentEnd;n>=alignmentEnd-extraCheckLen;n--){
+            //
+            char refNuc = tgr._getGenomeNuc(n,true);
+            int idx = sam.getReadPositionAtReferencePosition(n)-1;
+            if(idx>=0 && idx<sam.getReadLength()){
+                char readNuc = sam.getReadString().charAt(idx);
+                if(!equalnuc(refNuc,readNuc)){
+                    miscount++;
+                }
+            }
+        }
+
+        return miscount;
+    }
+
+    private static boolean equalnuc(char refNuc,char readNuc) {
+        return Character.toUpperCase(refNuc) == Character.toUpperCase(readNuc);
+    }
+
+    private static boolean isClipped(CigarElement ce){
```

Comment:
It looks like an unused function.
I think it is better to delete it if it is not needed.